### PR TITLE
[1.8][external assets] accept `AssetSpec`s in `materialize`

### PR DESCRIPTION
**Pull request by sryza**
## Summary & Motivation

`SourceAsset`s can currently be passed to the `assets` argument of `materialize`. This allows metadata about an upstream asset to be made available during the execution of an asset downstream of it – e.g. to grab the upstream `PartitionsDefinition` to resolve the upstream partitions that the execution depends on.

This PR enables the same with `AssetSpec`s, given the impending deprecation of `SourceAsset`.

This PR depends on `Definitions` accepting `AssetSpec`s (https://github.com/dagster-io/dagster/pull/22961).

### CODE REVIEW

File: python_modules/dagster/dagster_tests/asset_defs_tests/test_materialize.py [hunk 1]
```diff
@@ -151,6 +152,30 @@ def the_asset(the_source):
...
+def test_materialize_asset_specs():
+    class MyIOManager(IOManager):
+        def handle_output(self, context, obj):
+            pass
+
+        def load_input(self, context):
+            return 5
+
+    the_source = AssetSpec(
+        key=AssetKey(["the_source"]), metadata={"dagster/io_manager_key": "my_io_manager"}
```
Comment (sryza) [inline python_modules/dagster/dagster_tests/asset_defs_tests/test_materialize.py:164]: this likely merits a discussion

Comment (sryza) [reply python_modules/dagster/dagster_tests/asset_defs_tests/test_materialize.py:164]: Some background here: https://github.com/dagster-io/internal/discussions/10536#discussioncomment-10036256

**Review by schrockn — COMMENTED** [review]
Can you add a test case where someone adds an `AssetSpec`  that is already added in an `AssetsDefinition`? I think understand that error condition is key here. Overall this makes sense.

**Review by schrockn — CHANGES_REQUESTED** [review]
Test case for conflictig spec and assetsdefinition.

**Comment by sryza** [comment]
👍 @schrockn added that test case: https://github.com/dagster-io/dagster/pull/23216/files#diff-10e165d419df4e8aa25f64b26f2445f4f3338e0ba5a3d37fdc1eaa2a2c946c9eR180

**Review by schrockn — APPROVED** [review]
Cool

**Pull request merged**